File: python/examples/macmpec_lcqp.py
import h5py
import numpy as np
import rcqp
import json
import logging
import os


def load_problem(path):
    with h5py.File(path, "r") as f:
        return {k: np.asarray(f[k]).T for k in f.keys()}


def solve_macmpec_lcqp(name):
    data = load_problem(os.path.join(os.path.dirname(__file__), f"../../examples/data/macmpec/h5/{name}.h5"))
    logger.info(f"{data['n_eq']} equality constraints, {data['n_ineq']} inequality constraints, {data['n_comp']} complementarity constraints")
    
    H,      q,      f0     = data["H"],      data["q"],      float(data["f0"].item())
    J_eq,   b_eq           = data["J_eq"],   data["b_eq"]
    J_ineq, b_ineq         = data["J_ineq"], data["b_ineq"]
    J_comp, b_comp         = data["J_comp"], data["b_comp"]

    obj = lambda z: 0.5 * z.T @ H @ z + q.T @ z + f0

    solver  = rcqp.Solver()
    problem = rcqp.Problem(H, q, f0,
                           J_eq,   b_eq,
                           J_ineq, b_ineq,
                           J_comp, b_comp)
    solver.set_problem(problem)
    assert solver.solve(rcqp.SolverOptions()), "Failed to solve problem"

    logger.info(f"\tObjective value: {obj(solver.get_workspace().z)}")


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    logger = logging.getLogger(__name__)

    json_path = os.path.join(os.path.dirname(__file__), "../../examples/data/macmpec/macmpec_lcqp.json")
    with open(json_path, "r") as f:
        json_data = json.load(f)

    for key in sorted(json_data.keys()):
        logger.info(f"Solving MacMPEC problem: {key}")
        solve_macmpec_lcqp(key)

File: julia/examples/macmpec_lcqp.jl
# Run all of the MacMPEC QPCC problems with Marble and extract the results

using HDF5
using JSON
using Marble
using AmplNLReader
using MPCCModels

## Function to solve an individual problem
function mpcc_from_ampl(ampl::AmplNLReader.AmplModel)
    # First we find the nonzero elements in the cvar vector:
    # cvar = 0 → normal nonlinear constraint
    # cvar ≠ 0 → complementarity with the cvar-th variable
    ind_ccc2 = findall(!iszero, ampl.meta.cvar)
    # Then we store the corresponding variables
    ind_vcc1 = ampl.meta.cvar[ind_ccc2]
    return MPCCModels.MPCCModelVarCon(ampl, ind_vcc1, ind_ccc2)
end

function solve_macmpec_lcqp(name)
    nl_path = joinpath(@__DIR__, "..", "..", "examples", "data", "macmpec", "ampl_nl", name * ".nl")
    ampl_model = AmplModel(nl_path)
    mpcc_model = mpcc_from_ampl(ampl_model)
    data = from_mpcc(mpcc_model)

    # Instantiate probem
    problem = Marble.Problem(data.Q, data.q, data.c0, data.J_eq, data.b_eq, data.J_ineq, data.b_ineq, data.L, data.l, data.R, data.r)

    # Create an instance of Marble and solve the problem
    solver = Marble.Solver()
    Marble.set_problem!(solver, problem)

    opts = Marble.SolverOptions()
    Marble.verbosity!(opts, true)

    result = Marble.solve!(solver, opts)
    @assert Marble.converged(result) "Solver failed to converge on problem $name"

    # @info("Solving MacMPEC problem: $name",
    #     n_eq = Marble.n_eq(problem),
    #     n_ineq = Marble.n_ineq(problem),
    #     n_comp = Marble.n_comp(problem),
    #     iterations = Marble.iterations(result),
    #     objective = Marble.obj(data, Marble.z(result)),
    # )
end

## Run the function for MacMPEC QPCC problems

json_path = joinpath(@__DIR__, "../../examples/data/macmpec/macmpec_lcqp.json")
json_data = JSON.parsefile(json_path)

for key in sort(collect(keys(json_data)))[end:end]
    solve_macmpec_lcqp(key)
end

@info("All MacMPEC QPCC problems solved successfully")